File: Game.cs
using System.Numerics;
using Raylib_cs;

namespace Imperialism;

class Game {
    private Map _map;
    private CameraController _cameraController;
    private Camera2D _camera;
    private Player _player;

    public Game() {
        Raylib.InitWindow(1280, 720, "Imperialism");
        _map = new();
        _cameraController = new();
        _camera = new(new Vector2(Settings.TotalWidth / 2, Settings.TotalHeight / 2), _cameraController.Position, 0, _cameraController.Zoom);
        _player = new("Player 1");

        _map.RedrawMap();
        EventLoop();

        Raylib.CloseWindow();
    }

    private void EventLoop() {
        while (!Raylib.WindowShouldClose()) {
            float dt = Raylib.GetFrameTime();
            _cameraController.Update(dt);
            _player.Update(dt, _camera);
            _camera.Target = _cameraController.Position;
            _camera.Zoom = _cameraController.Zoom;

            Raylib.BeginDrawing();
            Raylib.ClearBackground(Color.White);

            Raylib.BeginMode2D(_camera);
            
            _map.Draw();
            _player.Draw(); 
            
            Raylib.EndMode2D();
            
            _player.DrawHUD();



            Raylib.EndDrawing();
        }
    }
}

File: game.go
package main

import rl "github.com/gen2brain/raylib-go/raylib"

var (
	width     int          = 1280
	height    int          = 720
	screenRec rl.Rectangle = rl.NewRectangle(0, 0, float32(width), float32(height))
	dt        float32      = 0
)

const (
	cameraSpeed     = 500
	cameraZoomSpeed = 0.125
)

type Game struct {
	Atlas  Atlas
	Camera rl.Camera2D
}

func NewGame() Game {
	return Game{
		Atlas:  NewAtlas(),
		Camera: rl.NewCamera2D(rl.NewVector2(-float32(width)/2, -float32(height)/2), rl.Vector2Zero(), 0, 1),
		// Camera: rl.NewCamera2D(rl.Vector2Zero(), rl.Vector2Zero(), 0, 1),
	}
}

func (game *Game) Update() {
	width = rl.GetScreenWidth()
	height = rl.GetScreenHeight()
	screenRec = rl.NewRectangle(0, 0, float32(width), float32(height))

	dt = rl.GetFrameTime()

	game.Camera.Zoom += game.Camera.Zoom * cameraZoomSpeed * rl.GetMouseWheelMoveV().Y

	// Control the camera
	game.Camera.Offset.X = float32(width) / 2
	game.Camera.Offset.Y = float32(height) / 2

	var speed float32 = cameraSpeed

	if SpeedModifier() {
		speed *= 5
	}

	if Up() {
		game.Camera.Target.Y -= speed / game.Camera.Zoom * dt
	}
	if Left() {
		game.Camera.Target.X -= speed / game.Camera.Zoom * dt
	}
	if Down() {
		game.Camera.Target.Y += speed / game.Camera.Zoom * dt
	}
	if Right() {
		game.Camera.Target.X += speed / game.Camera.Zoom * dt
	}

	game.Atlas.Update()
}

func (game *Game) Draw() {
	rl.ClearBackground(rl.RayWhite)

	rl.BeginMode2D(game.Camera)

	rl.DrawCircleV(rl.Vector2Zero(), 25, rl.Red)
	game.Atlas.Draw()

	rl.EndMode2D()
}
